### PR TITLE
Split the overlay into peer-2-peer and application layers.

### PR DESCRIPTION
**Pull request by gregtatcam**
## High Level Overview of Change

Split PeerImp and OverlayImpl into p2p layer and application layer -
P2PeerImp, P2POverlayImpl and PeerImp, P2POverlayImpl. p2p layer
is responsible for peer discovery, establishing and
maintaining connections between the peers, and sending and receiving
protocol messages. p2p layer implements only p2p methods
of Peer and Overlay classes. Application layer PeerImp and OverlayImpl
inherit from p2p layer classes and implement remaining methods in
Peer and Overlay. Application layer handles protocol message handling
including broadcast, relay, and squelching.

### Context of Change

This is a refactoring task. It doesn't add or delete existing functionality.
It facilitates testing of the overlay functions by removing the overlay
dependency on the application functionality.

### Type of Change

- [ x] Refactor (non-breaking change that only restructures code)

## Test Plan

Two new manual tests overlay_net and overlay_xrpl are added. overlay_net test the overlay creation
of 5 virtual nodes. overlay_xrpl is more of a simulation test. It creates the overlay of complete
XRPL network as described by and adjacency matrix generated from crawling XRPL network.

### CODE REVIEW

File: src/test/overlay/overlay_test.cpp [hunk 1]
```diff
@@ -0,0 +1,1403 @@
...
+        testXRPLOverlay();
+    }
+};
+
+BEAST_DEFINE_TESTSUITE_MANUAL(overlay_net, overlay, ripple);
+BEAST_DEFINE_TESTSUITE_MANUAL(overlay_xrpl, overlay, ripple);
+
+}  // namespace test
+
+}  // namespace ripple
```
Comment (HowardHinnant) [inline src/test/overlay/overlay_test.cpp:1403]: Needs newline at end of file.

File: src/test/overlay/overlay_test.cpp [hunk 2]
```diff
@@ -0,0 +1,1403 @@
...
+
+    void
+    run() override
+    {
+        testXRPLOverlay();
+    }
+};
+
+BEAST_DEFINE_TESTSUITE_MANUAL(overlay_net, overlay, ripple);
+BEAST_DEFINE_TESTSUITE_MANUAL(overlay_xrpl, overlay, ripple);
```
Comment (HowardHinnant) [inline src/test/overlay/overlay_test.cpp:1399]: These two tests fail for me:
```
Jade-5:~/Development/build> rippled --unittest=overlay_net
ripple.overlay.overlay_net Overlay
ERR:Server0 Bind port '' failed:Can't assign requested address             
#1 failed: unhandled exception: std::exception
313ms, 1 suite, 1 case, 1 test total, 1 failure
Jade-5:~/Development/build> rippled --unittest=overlay_xrpl
ripple.overlay.overlay_xrpl XRPLOverlay
#1 failed: adjacency matrix must be provided
102ms, 1 suite, 1 case, 1 test total, 1 failure
```

File: src/ripple/overlay/impl/P2PConfig.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+class PublicKey;
+class SecretKey;
+
+/** P2P required configuration properties.
+ */
+class P2PConfig
+{
+public:
+    P2PConfig() = default;
+    virtual ~P2PConfig() = default;
```
Comment (HowardHinnant) [inline src/ripple/overlay/impl/P2PConfig.h:43]: Should delete the copy members:
```c++
    P2PConfig(P2PConfig const&) = delete;
    P2PConfig& operator=(P2PConfig const&) = delete;
```

File: src/ripple/overlay/impl/P2PConfigImpl.h [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+#define RIPPLE_OVERLAY_P2PCONFIGIMPL_H_INCLUDED
+
+namespace ripple {
+
+/** Get P2P required configuration properties from the Application */
+class P2PConfigImpl : public P2PConfig
+{
+    Application& app_;
+
+public:
```
Comment (HowardHinnant) [inline src/ripple/overlay/impl/P2PConfigImpl.h:37]: Should delete the copy members:
```c++
    P2PConfigImpl(P2PConfigImpl const&) = delete;
    P2PConfigImpl& operator=(P2PConfigImpl const&) = delete;
```

File: src/ripple/overlay/impl/P2POverlayImpl.h [hunk 5]
```diff
@@ -0,0 +1,1160 @@
...
+     * are destroyed.
+     */
+    class Child
+    {
+    protected:
+        P2POverlayImpl<PeerImp_t>& overlay_;
+
+        explicit Child(P2POverlayImpl<PeerImp_t>& overlay);
+
+        virtual ~Child();
```
Comment (HowardHinnant) [inline src/ripple/overlay/impl/P2POverlayImpl.h:135]: Should delete the copy members:
```c++
        Child(Child const&) = delete;
        Child& operator=(Child const&) = delete;
```

File: src/ripple/overlay/impl/P2POverlayImpl.h [hunk 6]
```diff
@@ -0,0 +1,1160 @@
...
+        stop() = 0;
+    };
+
+private:
+    /** Once a second overlay timer */
+    struct Timer : Child, std::enable_shared_from_this<Timer>
+    {
+        boost::asio::basic_waitable_timer<clock_type> timer_;
+        bool stopping_{false};
+
```
Comment (HowardHinnant) [inline src/ripple/overlay/impl/P2POverlayImpl.h:148]: Should delete the copy members:
```c++
        Timer(Timer const&) = delete;
        Timer& operator=(Timer const&) = delete;
```

**Review by HowardHinnant — CHANGES_REQUESTED** [review]

File: src/test/overlay/overlay_test.cpp (hunk 2 above)
Comment (gregtatcam) [reply src/test/overlay/overlay_test.cpp:1399]: Ip addresses have to be pre-configured for both tests. I have comments that explain how to run the test 
 and configure IP's: https://github.com/gregtatcam/rippled/blob/overlay-restr-partial1-templ/src/test/overlay/overlay_test.cpp#L63 and https://github.com/gregtatcam/rippled/blob/overlay-restr-partial1-templ/src/test/overlay/overlay_test.cpp#L1029. I can test for the configured IP's and provide a better error message if not configured and a better message if the adjacency matrix is missing. Does this work?

File: src/ripple/overlay/impl/P2PConfig.h (hunk 3 above)
Comment (gregtatcam) [reply src/ripple/overlay/impl/P2PConfig.h:43]: Fixed

File: src/test/overlay/overlay_test.cpp (hunk 2 above)
Comment (HowardHinnant) [reply src/test/overlay/overlay_test.cpp:1465]: Your response is good enough.  I didn't read the comments.  The way you have it sounds good to me, thanks.

**Review by HowardHinnant — APPROVED** [review]
Looks good!

**Comment by intelliot** [comment]
(notes: significant rearchitecture of the code; @nbougalis will review)

**Comment by intelliot** [comment]
@nbougalis : Do you still intend to review this? If not, could you suggest 1-2 other reviewers?

@gregtatcam : If you still think this is worth doing, could you bring it up-to-date with `develop` (i.e. resolve conflicts)? 

Thanks!

**Comment by gregtatcam** [comment]
> @nbougalis : Do you still intend to review this? If not, could you suggest 1-2 other reviewers?
> 
> @gregtatcam : If you still think this is worth doing, could you bring it up-to-date with `develop` (i.e. resolve conflicts)?
> 
> Thanks!

It's a substantial overlay's refactoring. I don't think the benefits outweigh the risks at this point.

**Comment by intelliot** [comment]
@gregtatcam  Okay. Can you close this pull request?

**Pull request closed without merging**